client/ayon_core/hosts/max/plugins/publish/validate_instance_in_context.py

```diff
@@ -0,0 +1,99 @@
+from pymxs import runtime as rt
+
+
+class ValidateInstanceInContext(pyblish.api.InstancePlugin,
+                                OptionalPyblishPluginMixin):
+    """Validator to check if instance asset match context asset.
+
+    When working in per-shot style you always publish data in context of
+    current asset (shot). This validator checks if this is so. It is optional
+    so it can be disabled when needed.
+
+    Action on this validator will select invalid instances.
+    """
+    order = ValidateContentsOrder
+    label = "Instance in same Context"
+    optional = True
+    hosts = ["max"]
+    actions = [SelectInvalidAction, RepairAction]
+
+    def process(self, instance):
+        if not self.is_active(instance.data):
+            return
+
+        instance_node = rt.getNodeByName(instance.data.get(
+            "instance_node", ""))
+        if not instance_node:
+            return
+        asset = rt.getUserProp(instance_node, "folderPath")
+        context_asset = instance.context.data["folderPath"]
+        task = rt.getUserProp(instance_node, "task")
+        context_task = instance.context.data["task"]
+        invalid = []
+        if asset != context_asset:
+            invalid.append(
+                "Instance '{}' publishes to different asset than current "
+                "context: {}. Current context: {}".format(
+                    instance.name, asset, context_asset
+                )
+            )
+        if task != context_task:
+            invalid.append(
+                "Instance '{}' publishes to different task than current "
+                "context: {}. Current context: {}".format(
+                    instance.name, task, context_task
+                )
+            )
+
+        if invalid:
+            message = "\n".join(invalid)
+            raise PublishValidationError(
+                message=message,
+                description=(
+                    "## Publishing to a different asset\n"
+                    "There are publish instances present which are publishing "
+                    "into a different asset or task than your current context.\n\n"
+                    "Usually this is not what you want but there can be cases "
+                    "where you might want to publish into another asset or "
+                    "shot. If that's the case you can disable the validation "
+                    "on the instance to ignore it."
+                )
+            )
+
+    @classmethod
+    def get_invalid(cls, instance):
+        invalid = []
+        node = rt.getNodeByName(instance.data["instance_node"])
+        asset = rt.getUserProp(node, "folderPath")
+        context_asset = instance.context.data["folderPath"]
+        if asset != context_asset:
+            invalid.append(node)
+        task = rt.getUserProp(node, "task")
+        context_task = instance.context.data["task"]
+        if task != context_task:
+            invalid.append(node)
+
+        return invalid
```

Comment:
Usually it's the case that `get_invalid` logic is re-used across the plug-in. Currently it isn't. It's not a huge issue, but basically means that `selectinvalidaction` currently relies on different checking logic than `process()` does which could mean more maintenance down the line.

Other than that, overall logic looks good. Can't test myself, not a Max user :)

Reply:
Oh I did want to add! :)

I think with the NEW publisher we can actually implement this DCC agnostic as a global plug-in.

We can just check `instance.data["folderPath"]` against the `context.data["folderPath"]` and validate that way.
Then to repair, we would just tweak the values in `CreateContext` on the `CreatedInstance` instances, change the values, trigger save on the CreateContext and voila.

Reply:
And then I remembered. I already did so, but just implemetned it for Houdini: https://github.com/ynput/OpenPype/blob/b4d5a7465e95992e1ae07894f62e189e3559b67d/openpype/hosts/houdini/plugins/publish/validate_instance_in_context.py#L14-L13

Pretty sure that would work for Max too then? (except for the SelectROPAction)

Reply:
> And then I remembered. I already did so, but just implemetned it for Houdini: https://github.com/ynput/OpenPype/blob/b4d5a7465e95992e1ae07894f62e189e3559b67d/openpype/hosts/houdini/plugins/publish/validate_instance_in_context.py#L14-L13
> 
> Pretty sure that would work for Max too then? (except for the SelectROPAction)

yes, pretty sure it will work. It's just that we need to have separate functions for invalid actions.